core/Enclave/enclave_hsm.cpp

```diff
@@ -101,696 +93,304 @@ static const sgx_ec256_public_t g_sp_pub_key = {
-
-static uint32_t sgx_calc_gcm_data_size(const uint32_t aad_size, const uint32_t plaintext_size)
-{
-    if (aad_size > UINT32_MAX - sizeof(sgx_aes_gcm_data_ex_t))
-        return UINT32_MAX;
-
-    if (plaintext_size > UINT32_MAX - sizeof(sgx_aes_gcm_data_ex_t))
-        return UINT32_MAX;
-
-    if (aad_size > UINT32_MAX - plaintext_size)
-        return UINT32_MAX;
-
-    if (sizeof(sgx_aes_gcm_data_ex_t) > UINT32_MAX - plaintext_size - aad_size)
-        return UINT32_MAX;
-
-    return (aad_size + plaintext_size + sizeof(sgx_aes_gcm_data_ex_t));
-}
-
-static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_data)
-{
-    if (NULL == gcm_data)
-        return UINT32_MAX;
-
-    return gcm_data->ciphertext_size;
-}
-
-
-static sgx_status_t sgx_gcm_encrypt(const sgx_aes_gcm_128bit_key_t *key,
-                                    const uint32_t plaintext_size, const uint8_t *plaintext,
-                                    const uint32_t aad_size, const uint8_t *aad,
-                                    const uint32_t gcm_data_size, sgx_aes_gcm_data_ex_t *gcm_data)
+sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_len)
 {
-    if (NULL == gcm_data || gcm_data_size < sgx_calc_gcm_data_size(aad_size, plaintext_size))
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    if (plaintext_size == 0 || NULL == plaintext)
-        return SGX_ERROR_INVALID_PARAMETER;
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
 
-    if (plaintext_size > UINT32_MAX - aad_size)
+    if (cmk == NULL || cmk->metadata.origin != EH_INTERNAL_KEY) {
         return SGX_ERROR_INVALID_PARAMETER;
-
-    uint32_t real_aad_size = aad_size;
-    if (NULL == aad)
-        real_aad_size = 0;
-
-    sgx_status_t ret = sgx_read_rand(gcm_data->iv, sizeof(gcm_data->iv));
-    if (ret != SGX_SUCCESS) {
-        printf("error generating iv.\n");
-        return ret;
     }
 
-    ret = sgx_rijndael128GCM_encrypt(key,
-                                     plaintext, plaintext_size,
-                                     gcm_data->payload,
-                                     gcm_data->iv, sizeof(gcm_data->iv),
-                                     &(gcm_data->payload[gcm_data->ciphertext_size]), real_aad_size,
-                                     reinterpret_cast<uint8_t (*)[16]>(gcm_data->mac));
-    if (SGX_SUCCESS != ret) {
-        printf("gcm encrypting failed.\n");
-    }
-    else {
-        gcm_data->ciphertext_size = plaintext_size;
-        gcm_data->aad_size = real_aad_size;
-    }
-
-    return ret;
-}
-
-static sgx_status_t sgx_gcm_decrypt(const sgx_aes_gcm_128bit_key_t *key,
-                                    uint32_t *plaintext_size, uint8_t *plaintext,
-                                    const sgx_aes_gcm_data_ex_t *gcm_data)
-{
-    if (NULL == gcm_data || NULL == plaintext || NULL == *plaintext_size
-                 || *plaintext_size < sgx_get_gcm_ciphertext_size(gcm_data))
+    if (cmk->metadata.keyspec >= INVALID_VALUE)
```

Comment:
how about the case <0, seems it's useless, because you already have siwtch..case to handle it.

Reply:
got it